### PR TITLE
Google Analytics V4 API uptake.

### PR DESCRIPTION
**Pull request by hidekoji**
# Description

Support Google Analytics V4 API uptake.

# Checklist
Make sure you have performed the following items before submitting this pull request.
If not, please describe the reason.  

- [ ] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/google_analytics.R [hunk 1]
```diff
@@ -19,10 +19,42 @@ getGoogleProfile <- function(tokenFileId = ""){
       data.frame(googleAnalyticsR::ga_view(accountId, webPropertyId, viewId)) %>% dplyr::filter(effective == "READ_AND_ANALYZE") %>% dplyr::select(id, accountId, webPropertyId, timezone)
     })
     # Join the timezone column to the original data frame.
-    df <- df %>% dplyr::left_join(newdf, by = c("accountId" = "accountId", "webPropertyId" = "webPropertyId", "viewId" = "id"))
+    df <- df %>% dplyr::left_join(newdf, by = c("accountId" = "accountId", "webPropertyId" = "webPropertyId", "viewId" = "id")) %>% dplyr::select(accountId, accountName, webPropertyId, webPropertyName, viewId, viewName, timezone)
```
Comment (hideaki) [inline R/google_analytics.R:22]: Add comment on why we have select here.

File: R/google_analytics.R [hunk 2]
```diff
@@ -19,10 +19,42 @@ getGoogleProfile <- function(tokenFileId = ""){
...
+  }
+  # get V4 Account
+  v4df <- googleAnalyticsR::ga_account_list("ga4")
+  if (nrow(v4df) > 0) {
+    v4newdf <- purrr::map_dfr(v4df$accountId, function(id){
+      # account looks like accounts/123456 so get rid of accounts/ to get account ID part. As for property, it looks like properties/123345 so remove properties/ to get property id.
+      getGoogleAnalyticsV4Property(id) %>% mutate(accountId = stringr::str_replace(parent, "accounts/", ""), webPropertyId = stringr::str_replace(name, "properties/", ""))
```
Comment (hideaki) [inline R/google_analytics.R:29]: Add dplyr prefix to mutate.

File: R/google_analytics.R [hunk 3]
```diff
@@ -19,10 +19,42 @@ getGoogleProfile <- function(tokenFileId = ""){
...
+    })
+    v4newdf <- v4newdf %>% dplyr::distinct(accountId, name, .keep_all = T)
+    v4df <- v4df %>% dplyr::left_join(v4newdf, by = c("accountId" = "accountId", "propertyId" = "webPropertyId")) %>% mutate(viewId = "", viewName = "") %>% dplyr::select(accountId, account_name, propertyId, property_name, viewId, viewName, timeZone) %>% dplyr::rename(webPropertyId = propertyId, webPropertyName = property_name, accountName = account_name, timezone = timeZone)
```
Comment (hideaki) [inline R/google_analytics.R:32]: mutate to align columns before bind_rows is probably unnecessary.

File: R/google_analytics.R [hunk 4]
```diff
@@ -60,6 +92,7 @@ getGoogleAnalytics <- function(tableId, lastNDays = 30, dimensions, metrics, tok
     segments = NULL
   }
   token <- getGoogleTokenForAnalytics(tokenFileId)
+  googleAuthR::gar_auth(token = token, skip_fetch = TRUE)
```
Comment (hideaki) [inline R/google_analytics.R:95]: Move this under isV4 if statement.

File: R/google_analytics.R [hunk 5]
```diff
@@ -186,12 +233,19 @@ getGoogleAnalytics <- function(tableId, lastNDays = 30, dimensions, metrics, tok
...
+
   if("dateHourMinute" %in% colnames(ga.data)){
     # modify date column to POSIXct object from integer like 202001210000
     loadNamespace("lubridate")
     ga.data <- ga.data %>% dplyr::mutate( dateHourMinute = lubridate::ymd_hms(dateHourMinute, truncated = 1) )
   }
 
+  if("firstSessionDate" %in% colnames(ga.data)){
+    # modify date column to Date object from integer like 20140101
+    loadNamespace("lubridate")
```
Comment (hideaki) [inline R/google_analytics.R:245]: loadNamespace here is unnecessary.

File: R/google_analytics.R [hunk 6]
```diff
@@ -186,12 +233,19 @@ getGoogleAnalytics <- function(tableId, lastNDays = 30, dimensions, metrics, tok
     ga.data <- ga.data %>% dplyr::mutate( dateHour = lubridate::ymd_h(dateHour) )
   }
 
+
   if("dateHourMinute" %in% colnames(ga.data)){
     # modify date column to POSIXct object from integer like 202001210000
     loadNamespace("lubridate")
     ga.data <- ga.data %>% dplyr::mutate( dateHourMinute = lubridate::ymd_hms(dateHourMinute, truncated = 1) )
```
Comment (hideaki) [inline R/google_analytics.R:240]: loadNamespace here is unnecessary.

File: R/google_analytics.R (hunk 1 above)
Comment (hidekoji) [reply R/google_analytics.R:22]: Added comment.

File: R/google_analytics.R (hunk 2 above)
Comment (hidekoji) [reply R/google_analytics.R]: Added dplyr prefix/

File: R/google_analytics.R (hunk 3 above)
Comment (hidekoji) [reply R/google_analytics.R]: removed unnecessary murate.

File: R/google_analytics.R (hunk 4 above)
Comment (hidekoji) [reply R/google_analytics.R]: moved the code inside isV4 if block.

File: R/google_analytics.R (hunk 5 above)
Comment (hidekoji) [reply R/google_analytics.R]: Removed it.

File: R/google_analytics.R (hunk 6 above)
Comment (hidekoji) [reply R/google_analytics.R:237]: Removed it.

**Pull request merged**